File: math/calcualtor/basic.py


# Leetcode Q224. Basic Calculator
# Given a string s representing a valid expression, implement a basic calculator to evaluate it,
# and return the result of the evaluation.
# Note: You are not allowed to use any built-in function which evaluates strings
# as mathematical expressions, such as eval().


def calculate(s: str) -> int:

	signs = []
	result = [ 0 ]
	current = str()

	for char in s:
		match char:
			
			case "-":

				result[-1] += int(current or 0)
				current = "-"

			case "+":

				result[-1] += int(current or 0)
				current = str()

			case "(":

				result.append(0)
				signs.append(-1 if current else 1)
				current = str()

			case ")":

				result[-1] += int(current or 0)
				prev = result.pop() * signs.pop()
				result[-1] += prev
				current = str()

			case " ":	continue
			case _:		current += char

	return	result[-1] + int(current or 0)

File: math/calcualtor/basic.js


// Leetcode Q224. Basic Calculator
// Given a string s representing a valid expression, implement a basic calculator to evaluate it,
// and return the result of the evaluation.
// Note: You are not allowed to use any built-in function which evaluates strings
// as mathematical expressions, such as eval().


var calculate = function(s) {

	var signs = [];
	var result = [ 0 ];
	var current = "";
	var prev;

	for(var chr of s) {
		switch(chr) {

			case "-":

				result[result.length-1] += current.length ? parseInt(current) : 0;
				current = "-";
				break;

			case "+":

				result[result.length-1] += current.length ? parseInt(current) : 0;
				current = "";
				break;

			case "(":

				result.push(0);
				signs.push(current.length ? -1 : 1);
				current = "";
				break;

			case ")":

				result[result.length-1] += current.length ? parseInt(current) : 0;
				prev = result.pop() * signs.pop();
				result[result.length-1] += prev;
				current = "";
				break;

			case " ":

				continue;

			default:

				current += chr;
				break;
		}
	}	return	result[result.length-1] + (current.length ? parseInt(current) : 0);
}

